### ISSUE TITLE
DoS: Msg.Unpack uses untrusted lengths to allocate slices

### ISSUE THREAD

**Issue by lmb** [miekg/dns#609]
It is currently possible to make Msg.Unpack allocate several MiB of data per call, since lengths from the header are trusted. Any service using the library to parse packets from an untrusted source is therefore susceptible to DoS: an attacker can overwhelm the GC and force long GC pauses.

This was introduced in #281.

**Comment by miekg** [comment]
hmmm, can we just re-introduce:
~~~
dns.Question = make([]Question, 0, 1)
dns.Answer = make([]RR, 0, 10)		
￼dns.Ns = make([]RR, 0, 10)		
￼dns.Extra = make([]RR, 0, 10)
~~~

**Comment by miekg** [comment]
This might be good to fix and push out a new miekg/dns release. I also wanted to kick off coredns 1.0.2 but want to have this in first.

Do you have a PR in mind?

**Comment by lmb** [comment]
I can get to this tomorrow morning. I wanted to do roughly what you 
proposed, except not pre-allocating at all. The slice allocator always 
allocates powers of two anyways.

**Comment by miekg** [comment]
Thanks! Sgtm.

**Comment by andrewtj** [comment]
> an attacker can overwhelm the GC and force long GC pauses

If this is something we're concerned about then at least one RRData needs a rethink.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] DoS: Msg.Unpack use..." ]
>> an attacker can overwhelm the GC and force long GC pauses
>
>If this is something we're concerned about then at least one RRData needs a rethink.

should prolly do that. Was the idea we had backwards compatible?

**Comment by andrewtj** [comment]
On Thu, Dec 21, 2017 at 7:19 AM, Miek Gieben <notifications@github.com>
wrote:

> [ Quoting <notifications@github.com> in "Re: [miekg/dns] DoS: Msg.Unpack
> use..." ]
> >> an attacker can overwhelm the GC and force long GC pauses
> >
> >If this is something we're concerned about then at least one RRData needs
> a rethink.
>
> should prolly do that. Was the idea we had backwards compatible?
>
> —
>
No it wouldn't be a backward compatible change. I'm not sure I have a good
barometer for what's worth worrying about here, what with Go being a GC'd
language and this being a high-level(ish) DNS abstraction. With that said
and hopefully not putting anyone in the lurch here, both the type bitmap
and TXT representations could be a problem. I have crashed an echo server
by limiting its RAM footprint with ulimit and hammering it with messages
containing one of these types. It was a while ago and I didn't take notes,
but I think I curbed it to it ~10MB or so and sent a TXT payload which
cause it to croak within minutes.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] DoS: Msg.Unpack use..." ]
>On Thu, Dec 21, 2017 at 7:19 AM, Miek Gieben <notifications@github.com>
>wrote:
>
>> [ Quoting <notifications@github.com> in "Re: [miekg/dns] DoS: Msg.Unpack
>> use..." ]
>> >> an attacker can overwhelm the GC and force long GC pauses
>> >
>> >If this is something we're concerned about then at least one RRData needs
>> a rethink.
>>
>> should prolly do that. Was the idea we had backwards compatible?
>>
>> —
>>
>No it wouldn't be a backward compatible change. I'm not sure I have a good
>barometer for what's worth worrying about here, what with Go being a GC'd
>language and this being a high-level(ish) DNS abstraction. With that said
>and hopefully not putting anyone in the lurch here, both the type bitmap
>and TXT representations could be a problem. I have crashed an echo server
>by limiting its RAM footprint with ulimit and hammering it with messages
>containing one of these types. It was a while ago and I didn't take notes,
>but I think I curbed it to it ~10MB or so and sent a TXT payload which
>cause it to croak within minutes.

We should probably push through a change that fixes this.

/Miek

--
Miek Gieben

**Comment by andrewtj** [comment]
Okay. Giving it some more consideration: my laptop isn't representative of much and I didn't look into the behaviour I saw in any detail.

So it maybe that the cause isn't the RRDatas I picked but instead the volume of messages (#565 might fix that), and if it is the GC, the fix may be to tweak the collector in some way rather than changing the API.

I'm not going to dig into this one (I don't have the resources) but if anyone would like to [this](https://gist.github.com/andrewtj/209700b47932d9a1f4bed613ae9ca203) leaks slowly on my machine and uses relatively conservatively sized messages.
